lib/fabrik/VirtualHostManager.js

```diff
@@ -0,0 +1,168 @@
+    logger.info(`Creating binding '${binding.id}' for deployment '${deploymentName}', virtual host '${guid}'...`);
+    logger.info('+-> Binding parameters:', binding.parameters);
+    return this.getDeploymentIps(deploymentName)
+      .then(ips => this.agent.createCredentials(ips, guid, binding.parameters))
+      .tap(credentials => this.createBindingProperty(deploymentName, binding.id, _.set(binding, 'credentials', credentials)))
+      .tap(() => {
+        const bindCreds = _.cloneDeep(binding.credentials);
+        utils.maskSensitiveInfo(bindCreds);
+        logger.info(`+-> Created binding:${JSON.stringify(bindCreds)}`);
+      })
+      .catch(err => {
+        logger.error('+-> Failed to create binding');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  deleteBinding(deploymentName, guid, id) {
+    this.verifyFeatureSupport('credentials');
+    logger.info(`Deleting binding '${id}' for deployment '${deploymentName}' , virtual host '${guid}'...`);
+    return Promise
+      .all([
+        this.getDeploymentIps(deploymentName),
+        this.getBindingProperty(deploymentName, id)
+      ])
+      .spread((ips, binding) => this.agent.deleteCredentials(ips, guid, binding.credentials))
+      .then(() => this.deleteBindingProperty(deploymentName, id))
+      .tap(() => logger.info('+-> Deleted service binding'))
+      .catch(err => {
+        logger.error('+-> Failed to delete binding');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  saveVirtualHostInstanceToDeploymentRelation(deploymentName, id) {
+    return this.mapper
+      .createVirtualHostRelation(deploymentName, `virtual_host_instance-${id}`)
+      .catchThrow(BadRequest, new BadRequest(`Virtual host instance with ID ${id} already exists on deployment ${deploymentName}`));
+  }
+
+  deleteVirtualHostInstanceToDeploymentRelation(deploymentName, id) {
+    return this.mapper
+      .deleteVirtualHostRelation(`virtual_host_instance-${id}`)
+      .catchThrow(BadRequest, new BadRequest(`Virtual host instance with ID ${id} does not exists on deployment ${deploymentName}`));
+  }
+
+  getLockProperty(deploymentName) {
+    return this.director.getLockProperty(deploymentName);
+  }
+
+  createBindingProperty(deploymentName, id, value) {
+    return this.director
+      .createDeploymentProperty(deploymentName, `binding-${id}`, JSON.stringify(value))
+      .catchThrow(BadRequest, new ServiceBindingAlreadyExists(id));
+  }
+
+  getBindingProperty(deploymentName, id) {
+    return this.director
+      .getDeploymentProperty(deploymentName, `binding-${id}`)
+      .then(result => JSON.parse(result))
+      .catchThrow(NotFound, new ServiceBindingNotFound(id));
+  }
+
+  deleteBindingProperty(deploymentName, id) {
+    return this.director
+      .deleteDeploymentProperty(deploymentName, `binding-${id}`);
+  }
+
+  static get instanceConstructor() {
+    return VirtualHostInstance;
+  }
+
+  static load(plan) {
```

Comment:
code is the same as that in `DirectorManager` - can move this code to `BaseManager`. `DockerManager` anyways has a different implementation and that override can remain.